### PR TITLE
build ccoctl for Windows and MacOS

### PR DESCRIPTION
**Pull request by joelddiaz**
Introduce new Makefile targets to build for Windows and MacOS (amd64).

Update Dockerfile to build and save those binaries into final container
image.

xref: https://issues.redhat.com/browse/CCO-170

### CODE REVIEW

File: Dockerfile
Comment (akhil-rane) [inline Dockerfile]: Do we have linux binary in /usr/bin and mac/windows binary in /ccoctl/ ?

File: Dockerfile
Comment (akhil-rane) [inline Dockerfile:5]: Who sets this value?

File: Dockerfile
Comment (joelddiaz) [reply Dockerfile]: Yes. The linux binaries for all architectures will live in /usr/bin (but of course to get the arm64 linux binary you need to extract from an arm64 container image)
We just need a place to put these non-native binaries so that other tooling can extract and then sign them.

File: Dockerfile
Comment (joelddiaz) [reply Dockerfile:5]: The docs say this is automatically set https://docs.docker.com/engine/reference/builder/#automatic-platform-args-in-the-global-scope

And from my testing, buildah was setting it properly too https://github.com/containers/buildah/issues/1368

**Comment by joelddiaz** [comment]
> looks like no docs for this bit? /label docs-approved

If we get the signed binaries, maybe a one-liner about "you can get binaries for windows/mac from http://redhat.com/to-be-determined", but no major doc change

**Comment by jeana-redhat** [comment]
> > looks like no docs for this bit? /label docs-approved
> 
> If we get the signed binaries, maybe a one-liner about "you can get binaries for windows/mac from http://redhat.com/to-be-determined", but no major doc change

Sounds good! lmk if that ends up being needed

**Comment by lwan-wanglin** [comment]
@jianping-shu will take this card, thanks!
/assign @jianping-shu

**Comment by jianping-shu** [comment]
@joelddiaz @akhil-rane @lwan-wanglin Here are the test cases, PTAL!
[OCP-47064](https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitem?id=OCP-47064)
[OCP-47065](https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitem?id=OCP-47065)

**Comment by joelddiaz** [comment]
> @joelddiaz @akhil-rane @lwan-wanglin Here are the test cases, PTAL! [OCP-47064](https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitem?id=OCP-47064) [OCP-47065](https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitem?id=OCP-47065)

:+1: This looks correct for this PR. Eventually we want to sign/publish the binaries to a different location, but that hasn't happened yet and those test cases are the best way to test the results of building for MacOS and Windows binaries using only the release image.

**Comment by jianping-shu** [comment]
@joelddiaz Thanks for the info! After sign/publish steps are ready, I can update the test cases accordingly. 
/label qe-approved

**Comment by jeffnowicki** [comment]
@joelddiaz Can you comment on the status of the sign/publish steps? Is that nearly complete and if so, will the additional binaries be delivered for 4.10?

**Comment by joelddiaz** [comment]
> @joelddiaz Can you comment on the status of the sign/publish steps? Is that nearly complete and if so, will the additional binaries be delivered for 4.10?

Still waiting on px-approved label before this merges (cc @sferich888 ). Then there is additional work to put these binaries through the signing pipeline (unknown what these steps are as this team has never had to go through this process before).

**Comment by 2uasimojo** [comment]
This is being actively worked by @akhil-rane and @jianping-shu via https://issues.redhat.com/browse/CCO-170

/remove-lifecycle stale
/lifecycle frozen

**Comment by akhil-rane** [comment]
@sferich888 This needs PX approval

File: Dockerfile
Comment (yselkowitz) [reply Dockerfile:5]: I don't think this is supported in OSBS though.  But there is also no reason to limit the Mac and Windows binaries to just the amd64 payload though.

File: Dockerfile
Comment (yselkowitz) [inline Dockerfile:9]:
```suggestion
RUN make cross-build
```

Comment (yselkowitz) [inline Dockerfile:9]:
```suggestion
RUN make cross-build
```

File: Dockerfile
Comment (yselkowitz) [inline Dockerfile:15]:
```suggestion
COPY --from=builder /go/src/github.com/openshift/cloud-credential-operator/_output/bin/linux_amd64/ccoctl /usr/share/openshift/linux_amd64/ccoctl
COPY --from=builder /go/src/github.com/openshift/cloud-credential-operator/_output/bin/linux_arm64/ccoctl /usr/share/openshift/linux_arm64/ccoctl
COPY --from=builder /go/src/github.com/openshift/cloud-credential-operator/_output/bin/darwin_amd64/ccoctl /usr/share/openshift/mac/ccoctl
COPY --from=builder /go/src/github.com/openshift/cloud-credential-operator/_output/bin/darwin_arm64/ccoctl /usr/share/openshift/mac_arm64/ccoctl
COPY --from=builder /go/src/github.com/openshift/cloud-credential-operator/_output/bin/windows_amd64/ccoctl /usr/share/openshift/windows/ccoctl
```

Comment (yselkowitz) [inline Dockerfile:15]:
```suggestion
COPY --from=builder /go/src/github.com/openshift/cloud-credential-operator/_output/bin/linux_amd64/ccoctl /usr/share/openshift/linux_amd64/ccoctl
COPY --from=builder /go/src/github.com/openshift/cloud-credential-operator/_output/bin/linux_arm64/ccoctl /usr/share/openshift/linux_arm64/ccoctl
COPY --from=builder /go/src/github.com/openshift/cloud-credential-operator/_output/bin/darwin_amd64/ccoctl /usr/share/openshift/mac/ccoctl
COPY --from=builder /go/src/github.com/openshift/cloud-credential-operator/_output/bin/windows_amd64/ccoctl /usr/share/openshift/windows/ccoctl
```

File: Makefile
Comment (yselkowitz) [inline Makefile:143]:
```suggestion

cross-build-linux-amd64:
	+@GOOS=linux GOARCH=amd64 $(MAKE) --no-print-directory build GO_BUILD_PACKAGES:=./cmd/ccoctl GO_BUILD_BINDIR:=$(CROSS_BUILD_BINDIR)/linux_amd64
.PHONY: cross-build-linux-amd64

cross-build-linux-arm64:
	+@GOOS=linux GOARCH=arm64 $(MAKE) --no-print-directory build GO_BUILD_PACKAGES:=./cmd/ccoctl GO_BUILD_BINDIR:=$(CROSS_BUILD_BINDIR)/linux_arm64
.PHONY: cross-build-linux-arm64

cross-build-darwin-amd64:
	+@GOOS=darwin GOARCH=amd64 $(MAKE) --no-print-directory build GO_BUILD_PACKAGES:=./cmd/ccoctl GO_BUILD_FLAGS:="$(GO_BUILD_DARWIN)" GO_BUILD_BINDIR:=$(CROSS_BUILD_BINDIR)/darwin_amd64
.PHONY: cross-build-darwin-amd64

cross-build-darwin-arm64:
	+@GOOS=darwin GOARCH=arm64 $(MAKE) --no-print-directory build GO_BUILD_PACKAGES:=./cmd/ccoctl GO_BUILD_FLAGS:="$(GO_BUILD_DARWIN)" GO_BUILD_BINDIR:=$(CROSS_BUILD_BINDIR)/darwin_arm64
.PHONY: cross-build-darwin-arm64

cross-build-windows-amd64:
	+@GOOS=windows GOARCH=amd64 $(MAKE) --no-print-directory build GO_BUILD_PACKAGES:=./cmd/ccoctl GO_BUILD_FLAGS:="$(GO_BUILD_WINDOWS)" GO_BUILD_BINDIR:=$(CROSS_BUILD_BINDIR)/windows_amd64
.PHONY: cross-build-windows-amd64

cross-build: cross-build-linux-amd64 cross-build-linux-arm64 cross-build-darwin-amd64 cross-build-darwin-arm64 cross-build-windows-amd64
```

Comment (yselkowitz) [inline Makefile:143]:
```suggestion

cross-build-linux-amd64:
	+@GOOS=linux GOARCH=amd64 $(MAKE) --no-print-directory build GO_BUILD_PACKAGES:=./cmd/ccoctl GO_BUILD_BINDIR:=$(CROSS_BUILD_BINDIR)/linux_amd64
.PHONY: cross-build-linux-amd64

cross-build-linux-arm64:
	+@GOOS=linux GOARCH=arm64 $(MAKE) --no-print-directory build GO_BUILD_PACKAGES:=./cmd/ccoctl GO_BUILD_BINDIR:=$(CROSS_BUILD_BINDIR)/linux_arm64
.PHONY: cross-build-linux-arm64

cross-build-darwin-amd64:
	+@GOOS=darwin GOARCH=amd64 $(MAKE) --no-print-directory build GO_BUILD_PACKAGES:=./cmd/ccoctl GO_BUILD_FLAGS:="$(GO_BUILD_DARWIN)" GO_BUILD_BINDIR:=$(CROSS_BUILD_BINDIR)/darwin_amd64
.PHONY: cross-build-darwin-amd64

cross-build-windows-amd64:
	+@GOOS=windows GOARCH=amd64 $(MAKE) --no-print-directory build GO_BUILD_PACKAGES:=./cmd/ccoctl GO_BUILD_FLAGS:="$(GO_BUILD_WINDOWS)" GO_BUILD_BINDIR:=$(CROSS_BUILD_BINDIR)/windows_amd64
.PHONY: cross-build-windows-amd64

cross-build: cross-build-linux-amd64 cross-build-linux-arm64 cross-build-darwin-amd64 cross-build-windows-amd64
```

**Review by yselkowitz — CHANGES_REQUESTED** [review]
The following assumes support for the current public clouds only, which means only x86 and ARM.  If PowerVS (in development) is going to be supported herein, or if the existing IBM Cloud support covers PowerVS, then linux/ppc64le should also be included.

**Comment by akhil-rane** [comment]
> /lgtm cancel

we need to add ccoctl binaries for other platforms too.

**Comment by yselkowitz** [comment]
Reading more on `ccoctl`, it seems most uses are in conjunction with `openshift-install`.  Therefore, does it actually make sense to provide binaries for Windows, which does not have an installer binary?  If not, then we really only need to add linux/amd64, darwin/amd64, and darwin/arm64.  Together with the native linux binary already present (since 4.8), that would cover all the installer platforms for each payload.  Does that make sense?

**Comment by yselkowitz** [comment]
> Reading more on `ccoctl`, it seems most uses are in conjunction with `openshift-install`. Therefore, does it actually make sense to provide binaries for Windows, which does not have an installer binary? If not, then we really only need to add linux/amd64, darwin/amd64, and darwin/arm64. Together with the native linux binary already present (since 4.8), that would cover all the installer platforms for each payload. Does that make sense?

Ping?

**Comment by yselkowitz** [comment]
Yeah, taking a deeper look at the docs, while most cases are connected to installation/removal, there are indeed some use cases which do not seem to be in conjunction with the installer.  Windows makes that 4 (eventually 5) additional binaries, which in total weighs in at 332MB.  I strongly suggest adding a "ccoctl-artifacts" payload image to contain these without bloating the actual operator, which would involve a second Dockerfile in addition to the Makefile changes.

**Comment by yselkowitz** [comment]
Are you planning to move forward with this in 4.12?

**Pull request closed without merging**
